### PR TITLE
Roy/mempool subsystem

### PR DESCRIPTION
**Pull request by royster57**
Make the mempool a subsystem. We also add boilerplate code for the mempool's rpc server and pass a mempool handle to p2p. 
Note: Because of issues relating to the interface between mempool and chainstate, the mempool currently (temporarily) accepts both ChainstateInterface handle and a DummyMempoolChainstate, which reflects the trait that was used to mock the chainstate before it was written. The dummy will disappear once the mempool is properly linked to chainstate.

### CODE REVIEW

File: node/src/runner.rs [hunk 1]
```diff
@@ -64,13 +68,25 @@ pub async fn initialize(
         )?,
     );
 
+    // Mempool subsystem
+    let mempool = manager.add_subsystem(
+        "mempool",
+        mempool::make_mempool(
+            DummyMempoolChainState {},
+            chainstate.clone(),
+            mempool::pool::SystemClock {},
```
Comment (TheQuantumPhysicist) [inline node/src/runner.rs:77]: Perhaps we should move the TimeGetter from the Chainstate crate and then use it for mempool too. We don't want to end up with 5 different time structs. Not a huge priority though. A TODO in here will do for now.

**Review by TheQuantumPhysicist — APPROVED** [review]
Looks good, especially that it's mostly boilerplate. Many things can be improved later. Please allow the others to review the PR too before merging.

File: p2p/src/sync/mod.rs [hunk 2]
```diff
@@ -96,6 +97,10 @@ pub struct BlockSyncManager<T: NetworkingService> {
     /// Subsystem handle to Chainstate
     chainstate_handle: subsystem::Handle<Box<dyn chainstate_interface::ChainstateInterface>>,
 
+    /// Subsystem handle to Mempool
+    #[allow(unused)]
+    mempool_handle: subsystem::Handle<Box<dyn MempoolInterface<DummyMempoolChainState>>>,
+
```
Comment (altonen) [inline p2p/src/sync/mod.rs:103]: Is there a reason why syncing is given a handle to mempool?

**Review by altonen — APPROVED** [review]

File: p2p/src/sync/mod.rs (hunk 2 above)
Comment (TheQuantumPhysicist) [reply p2p/src/sync/mod.rs:103]: This doesn't seem to be appropriate. Please just pass the mempool handle to p2p, and let the p2p developer manage that internally. I'll make another comment on where to start removing.

File: p2p/src/lib.rs [hunk 3]
```diff
@@ -191,6 +193,7 @@ where
         }
         {
             let consensus_handle = consensus_handle.clone();
+            let mempool_handle = mempool_handle.clone();
```
Comment (TheQuantumPhysicist) [inline p2p/src/lib.rs:196]: This can be removed. Just keep the parameter for this function unused.

**Review by stanislav-tkach — APPROVED** [review]

**Review by iljakuklic — APPROVED** [review]
Good stuff. Just some minor comments

File: p2p/p2p-test-utils/src/lib.rs [hunk 4]
```diff
@@ -236,6 +237,24 @@ pub async fn start_chainstate(
     handle
 }
 
+pub async fn start_mempool(
+    chainstate: subsystem::Handle<Box<dyn ChainstateInterface>>,
+) -> subsystem::Handle<Box<dyn MempoolInterface<DummyMempoolChainState>>> {
+    let mut man = subsystem::Manager::new("TODO");
```
[the remark this answers is not in the document]
Comment (altonen) [reply p2p/p2p-test-utils/src/lib.rs:243]: It doesn't, I don't remember why it was named TODO

File: node/src/runner.rs (hunk 1 above)
Comment (royster57) [reply node/src/runner.rs:77]: Done, placed a TODO and opened an issue so we won't forget.

File: p2p/src/lib.rs [hunk 5]
```diff
@@ -239,6 +244,7 @@ pub async fn make_p2p<T>(
     chain_config: Arc<ChainConfig>,
     p2p_config: Arc<P2pConfig>,
     consensus_handle: subsystem::Handle<Box<dyn chainstate_interface::ChainstateInterface>>,
+    mempool_handle: subsystem::Handle<Box<dyn MempoolInterface<DummyMempoolChainState>>>,
```
[the remark this answers is not in the document]
Comment (royster57) [reply p2p/src/lib.rs:247]: Created a new issue for this.

File: mempool/src/pool.rs [hunk 6]
```diff
@@ -199,22 +200,48 @@ impl RollingFeeRate {
...
+    pub(crate) fn new(
+        chain_state: C,
+        chainstate_handle: H,
+        clock: T,
+        memory_usage_estimator: M,
+    ) -> Self {
+        Self {
+            store: MempoolStore::new(),
+            chain_state,
+            chainstate_handle,
```
[the remark this answers is not in the document]
Comment (royster57) [reply mempool/src/pool.rs:236]: When the mempool was first implemented, chainstate did not exist so the chainstate functionality required by the mempool was abstracted into a trait (see `ChainState` in pool.rs) and the mempool was implemented by accepting an object implementing this trait (and this was used for mocking in tests). Currently in the code the connection between this trait and the actual chainstate is nonexistent, but to fix this properly will require integrating Gosha's work on utxos. Thus for the sake of getting as much of the boilerplate out of the way as possible we now have the mempool hold these two handles, one of which should disappear with the integration with the utxo crate.

File: p2p/p2p-test-utils/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (royster57) [reply p2p/p2p-test-utils/src/lib.rs:243]: In any case this function will go now that I'm just passing a mempool handle to p2p and letting @altonen handle the handle. :)

File: p2p/src/lib.rs (hunk 3 above)
Comment (royster57) [reply p2p/src/lib.rs:196]: Done.

File: p2p/src/sync/mod.rs (hunk 2 above)
Comment (royster57) [reply p2p/src/sync/mod.rs:103]: Done.

File: mempool/src/pool.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (stanislav-tkach) [reply mempool/src/pool.rs:236]: Got it, thanks for the explanations!

File: p2p/p2p-test-utils/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (royster57) [reply p2p/p2p-test-utils/src/lib.rs:243]: This function has gone now that I'm just passing a mempool handle to p2p and letting Aaro handle it as he wishes :)

**Pull request merged**